src/libutil/file-system.cc

```diff
@@ -21,36 +21,43 @@ namespace fs = std::filesystem;
 
 namespace nix {
 
-Path absPath(PathView path, std::optional<PathView> dir, bool resolveSymlinks)
+#ifdef __GNUC__
+std::unique_ptr<char, FreeDeleter> getCwd()
 {
-    std::string scratch;
+    return std::unique_ptr<char, FreeDeleter> { getcwd(NULL, 0) };
+}
+#endif
 
-    if (path[0] != '/') {
-        // In this case we need to call `canonPath` on a newly-created
-        // string. We set `scratch` to that string first, and then set
-        // `path` to `scratch`. This ensures the newly-created string
-        // lives long enough for the call to `canonPath`, and allows us
-        // to just accept a `std::string_view`.
-        if (!dir) {
+
+Path absPath(PathView path, std::optional<PathView> dir, bool resolveSymlinks)
+{
+    /* In some branches case we create strings that, as r-values, would
+       not live long enough. We instead assign them to these variables
+       so they are only deallocated at the end of this function, after
+       the call to `canonPath` where they are used. */
+    std::string scratch0;
+    std::unique_ptr<char, FreeDeleter> scratch1;
+    [[maybe_unused]] char buf[PATH_MAX];
+
+    return canonPath(
+        path[0] == '/'
+            ? path
+            : (scratch0 = concatStrings(
+                dir
+                    ? *dir
+                    :
 #ifdef __GNU__
-            /* GNU (aka. GNU/Hurd) doesn't have any limitation on path
-               lengths and doesn't define `PATH_MAX'.  */
-            char *buf = getcwd(NULL, 0);
-            if (buf == NULL)
+                    /* GNU (aka. GNU/Hurd) doesn't have any limitation
+                       on path lengths and doesn't define `PATH_MAX'. */
+                    *(scratch1 = getCwd());
 #else
-            char buf[PATH_MAX];
-            if (!getcwd(buf, sizeof(buf)))
-#endif
-                throw SysError("cannot get cwd");
-            scratch = concatStrings(buf, "/", path);
-#ifdef __GNU__
-            free(buf);
+                    getcwd(buf, sizeof(buf))
+                        ? buf
+                        : throw SysError("cannot get cwd"),
```

Comment:
can't we encapsulate these two branches and the whole legacy mem management issue simply  into a function `std::string getCwd()` that does it all automagically (with the performance-question yes/no from earler in mind)? 

this seems like an unnecessary dance for a function whose concern is making paths absolute.